ProcessGroupGloo::allgather_into_tensor_coalesced

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118498
* #118986
* #118437
* #118911
* __->__ #118910

### Motivation
Despite our plan to reduce gloo usage, it is still being widely used as testing tool (in both the PyTorch CI and user tests) for code that only uses nccl in real world scenario. There's some coverage issues around all-gather and reduce-scatter variants, which are currently worked around in ugly ways (e.g. [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L216-L219) and [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L262-L272)). For native funcol I ran into the same issues but I'd rather just fix the coverage.

**I think it's reasonable to think of this as a fix rather than adding new features. This is orthogonal to the potential reduction of gloo usage**.

### This PR

This PR adds `ProcessGroupGloo::allgather_into_tensor_coalesced`.  This is very straightforward - `ProcessGroupGloo` already supports `allgather_coalesced`, to which we can funnel `allgather_into_tensor_coalesced`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
ProcessGroupGloo::allgather_into_tensor_coalesced

[ghstack-poisoned]

Commit:
Update on "ProcessGroupGloo::allgather_into_tensor_coalesced"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2153,6 +2153,21 @@ c10::intrusive_ptr<Work> ProcessGroupGloo::allgather_coalesced(
+    const AllgatherOptions& opts) {
+  TORCH_CHECK_EQ(outputs.size(), inputs.size());
+  std::vector<std::vector<at::Tensor>> output_lists(getSize());
+  for (auto& output : outputs) {
```
nit: I think this can be `const`.
```suggestion
  for (const auto& output : outputs) {
```

shuqiangzhang replied:
probably cannot do that because of std::move

awgu replied:
It looks like `chunks` would be a `vector` of new `Tensor` objects, so the `std::move` should not affect the original `output` tensor?
https://github.com/pytorch/pytorch/blob/65efbf078ced58240edc4abefff3cc9afc2b8a1f/aten/src/ATen/native/TensorShape.cpp#L925
I may be confused though 😭

awgu reviewed (commented):
For my understanding: why are the coverage issues referring to the coalesced collectives `allgather_into_tensor_coalesced` (and `reduce_scatter_tensor_coalesced`)? The two links in the PR description seem to refer to the non-coalesced version (`all_gather_into_tensor` and `reduce_scatter_tensor`) not being supported in gloo.

yifuwang commented:
@awgu sorry for the confusion. That's just general examples of how we are dealing with coverage issues. For examples of coalesced impls, see [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L342C5-L352) and [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L317-L339).

shuqiangzhang reviewed (approved):
LGTM

Commit:
Update on "ProcessGroupGloo::allgather_into_tensor_coalesced"


### Motivation
Despite our plan to reduce gloo usage, it is still being widely used as testing tool (in both the PyTorch CI and user tests) for code that only uses nccl in real world scenario. There's some coverage issues around all-gather and reduce-scatter variants, which are currently worked around in ugly ways (e.g. [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L216-L219) and [this](https://github.com/pytorch/pytorch/blob/b9e86bc93d0ad71c9a2f3a01c4a41ed5ee4b665f/torch/distributed/_functional_collectives_impl.py#L262-L272)). For native funcol I ran into the same issues but I'd rather just fix the coverage.

**I think it's reasonable to think of this as a fix rather than adding new features. This is orthogonal to the potential reduction of gloo usage**.

### This PR

This PR adds `ProcessGroupGloo::allgather_into_tensor_coalesced`.  This is very straightforward - `ProcessGroupGloo` already supports `allgather_coalesced`, to which we can funnel `allgather_into_tensor_coalesced`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
